### PR TITLE
Patch/GUM 57 error handling

### PR DESCRIPTION
**Pull request by R-Mielamud**

### CODE REVIEW

**Comment by liza-veis** [comment]
The problem with this solution is that in case of error thunkAction won't be stopped, it will get undefined as the result from a service and continue the action, so to prevent this we will need to check the result in each thunkAction and if it's undefined then reject the action. This isn't a very good solution, and we won't be able to determine if an error occurred or the request was successful but returned nothing. Services shouldn't be responsible for error handling. That's why I wrote that we need a middleware that will be connected to the store and will handle errors after thunkAction is rejected

**Comment by liza-veis** [comment]
And here we need to throw an error too, not return response

Screenshot 2022-03-26 at 11 48 37

**Comment by R-Mielamud** [comment]
> The problem with this solution is that in case of error thunkAction won't be stopped, it will get undefined as the result from a service and continue the action, so to prevent this we will need to check the result in each thunkAction and if it's undefined then reject the action. This isn't a very good solution, and we won't be able to determine if an error occurred or the request was successful but returned nothing. Services shouldn't be responsible for error handling. That's why I wrote that we need a middleware that will be connected to the store and will handle errors after thunkAction is rejected

We can have several http calls in one thunk, where some need notifying, some don't. Better we'll rethrow error after notifying  in http transport. Action will be automatically rejected.

**Comment by R-Mielamud** [comment]
> And here we need to throw an error too, not return response
> 
> Screenshot 2022-03-26 at 11 48 37

Fixed

File: src/common/types/http/http-options.ts [hunk 1]
```diff
@@ -6,6 +6,8 @@ type HttpOptions = {
   payload: BodyInit_ | null;
   hasAuth: boolean;
   query: Record<string, unknown>;
+  notifyError: boolean;
+  customErrorMessage: string;
```
Comment (liza-veis) [inline src/common/types/http/http-options.ts]: Maybe it's better to combine these properties into one config like this:
```
  type HttpErrorConfig = {
    notify?: boolean;
    message?: string;
  }
```

```suggestion
  errorConfig: HttpErrorConfig;
```

**Review by liza-veis — APPROVED** [review]

File: src/services/http/index.ts [hunk 2]
```diff
@@ -46,7 +53,7 @@ class Http {
     })
       .then(this.checkStatus.bind(this))
       .then(res => this.parseJSON<T>(res))
-      .catch(this.throwError);
+      .catch(err => this.handleError(err, notifyError, customErrorMessage));
```
Comment (katyakostikova) [inline src/services/http/index.ts:56]: Where do we have custom messages in code? I saw that you've replaced all try..catches in code, but you didn't pass any error messages. We have to add them ourselves or how does it work, could you explain please?

**Review by katyakostikova — APPROVED** [review]

File: src/services/http/index.ts (hunk 2 above)
Comment (R-Mielamud) [reply src/services/http/index.ts:56]: Now we don't have any custom messages in code. Error message is evaluated like this: `customMessage ?? messageThatCameFromServer ?? responseStatusMessage`

**Review by AleksandrTermenzhy — APPROVED** [review]

**Pull request merged**